### ISSUE TITLE
Critical bug in GMMHMM EM algorithm

### ISSUE THREAD

**Issue by michcio1234** [hmmlearn/hmmlearn#78]
### Introduction

For quite a long time I've been struggling with the problem of too small variances and decreasing LL during GMMHMM training. Finally, I've found rather a big problem.
(You can skip to one-before-last section to see what's wrong in the code.)
### What we have now

The `fit()` method in `_BaseHMM.py` basically works as follows:
1. Accumulate sufficient statistics (ss) for every training example (E step)
2. Do M step
3. Check convergence

In the case of GMMHMM, `_accumulate_sufficient_statistics()` method uses `sklearn.mixture.gmm`'s mechanisms to extract ss from training example. Namely, for each HMM state, it constructs a temporary GMM:

```
        tmp_gmm = GMM(g.n_components, covariance_type=g.covariance_type)
        tmp_gmm._set_covars(
            distribute_covar_matrix_to_match_covariance_type(
                np.eye(n_features), g.covariance_type,
                g.n_components))
```

And asks it to perform its M step:

```
        norm = tmp_gmm._do_mstep(X, gmm_posteriors, self.params)
```

Then, that GMM's parameters are converted to sufficient statistics. First and second moments (here, a little misleadingly, called `stats['means']` and `stats['covars']`) are multiplied by null moment (called `norm`):

```
        stats['norm'][state] += norm
        if 'm' in self.params:
            stats['means'][state] += tmp_gmm.means_ * norm[:, np.newaxis]
        if 'c' in self.params:
            if tmp_gmm.covariance_type == 'tied':
                stats['covars'][state] += tmp_gmm.covars_ * norm.sum()
            else:
                cvnorm = np.copy(norm)
                shape = np.ones(tmp_gmm.covars_.ndim)
                shape[0] = np.shape(tmp_gmm.covars_)[0]
                cvnorm.shape = shape
                stats['covars'][state] += tmp_gmm.covars_ * cvnorm
```
### And here is the problem.

While it is ok to do so for first order moments, doing that for second order moments is an error! Means and ovariances of a given mixture of `tmp_gmm` are estimated in M step according to equations like these:

```
new_means = first_moment / null_moment
new_covars = second_moment / null_moment - new_means^2
```

So to "roll back" to moments from means and covars, one has to do:

```
first_moment = new_means * null_moment  # this is done correctly in the current implementation
second_moment = (new_covars + new_means^2) * null_moment  # and this is NOT!
```
### What we can do

A hotfix for this problem would be:

In `hmm.py`, method `_accumulate_sufficient_statistics()`, line `610`:
change

```
stats['covars'][state] += tmp_gmm.covars_ * cvnorm
```

to

```
stats['covars'][state] += (tmp_gmm.covars_ + tmp_gmm.means_**2) * cvnorm
```

And in `hmm.py`, method `_do_mstep()`, line `636` (and respectively for full covariance matrix):
change

```
g.covars_ = (stats['covars'][state] + self.covars_prior) / cvnorm
```

to

```
g.covars_ = (stats['covars'][state] + self.covars_prior) / cvnorm - g.means_**2
```
### Final thoughts

All this leads to a doubt: do we really need to use `sklearn.mixture.gmm`'s methods? Wouldn't it be easier and clearer if just wrote the EM algorithm for GMMHMM?

Regards
Michal

**Comment by superbobry** [comment]
Hi Michal, thank you for a comprehensive bug report. 

I like the idea of freeing `GMMHMM` from `sklearn.mixture`, especially after seeing some of the API changes in scikit-learn/scikit-learn#4802. Feel free to submit a PR.

**Comment by Randyjp** [comment]
I'm confused about the current state of this issue, I know that if I install the version 2.0 using pip fitting a GMMHMM with a covariance other than 'diag' won't work. Currently, I cloned the masters branch and installed the package using this code. Should fitting a GMMHMM with a covar different than 'diag' work under those circumstances now? it says version 0.2.1 is  installed

**Comment by superbobry** [comment]
Hi Randy, the current master still contains the broken version of GMMHMM. 

There is an ongoing effort to fix this (see PR #107) and 0.2.1 would be out as soon as we get the new GMMHMM working.

**Comment by andreaschandra** [comment]
Hi all, should we do hotfix as @michcio1234 mentioned? if yes, i didn't find the line that he refered. my hmmlearn version is 0.2.0

**Comment by anntzer** [comment]
I believe this is fixed by #107; if not please request a reopen.
